soci/soci_index.go

```diff
@@ -698,12 +734,60 @@ func (b *IndexBuilder) buildSociLayer(ctx context.Context, desc ocispec.Descript
 	}
 	err = b.config.artifactsDb.WriteArtifactEntry(entry)
 	if err != nil {
-		return nil, err
+		return nil, nil, err
 	}
 
 	fmt.Printf("layer %s -> ztoc %s\n", desc.Digest, ztocDesc.Digest)
 	b.addSociLayerAnnotations(&desc, &ztocDesc, toc)
-	return &ztocDesc, err
+	return &ztocDesc, toc, nil
+}
+
+// buildPrefetchLayer builds and stores prefetch artifacts for layers that contain prefetch files
+func (b *IndexBuilder) buildPrefetchLayer(ctx context.Context, builtZtocs []*ztocWithLayer, layers []ocispec.Descriptor) []*ocispec.Descriptor {
+	if len(b.config.prefetchPaths) == 0 {
+		return nil
+	}
+
+	layerPrefetchMap := make(map[string][]string)
+	sortedZtocs := make([]*ztocWithLayer, len(builtZtocs))
+	copy(sortedZtocs, builtZtocs)
+	sort.Slice(sortedZtocs, func(i, j int) bool {
+		return sortedZtocs[i].index > sortedZtocs[j].index
+	})
+
+	for _, prefetchPath := range b.config.prefetchPaths {
+		for _, zwl := range sortedZtocs {
+			found := false
+			for _, metadata := range zwl.ztoc.TOC.FileMetadata {
+				if metadata.Name == prefetchPath {
+					layerPrefetchMap[zwl.layerDigest] = append(layerPrefetchMap[zwl.layerDigest], prefetchPath)
+					found = true
+					break
+				}
+			}
+			if found {
+				break // Found in this layer, stop searching other layers
+			}
+		}
+	}
+
+	if len(layerPrefetchMap) == 0 {
+		return nil
+	}
+
```

Comment:
Does it make sense to do a dedupe/merge pass here so that the final prefetch artifact is the minimum set of non-overlapping span ranges?